ledger/src/bank_forks_utils.rs

```diff
@@ -199,6 +203,35 @@ fn bank_forks_from_snapshot(
         process::exit(1);
     }
 
+    if snapshot_config.snapshot_from == SnapshotFrom::Dir {
+        match snapshot_utils::bank_from_latest_snapshot_dir(
+            &snapshot_config.bank_snapshots_dir,
+            genesis_config,
+            &process_options.runtime_config,
+            &account_paths,
+            process_options.debug_keys.clone(),
+            Some(&crate::builtins::get(
+                process_options.runtime_config.bpf_jit,
+            )),
+            process_options.account_indexes.clone(),
+            process_options.limit_load_slot_count_from_snapshot,
+            process_options.shrink_ratio,
+            process_options.verify_index,
+            process_options.accounts_db_config.clone(),
+            accounts_update_notifier.clone(),
+            exit,
+        ) {
+            Ok(bank) => {
+                let bank_forks = BankForks::new(bank);
+                return (Arc::new(RwLock::new(bank_forks)), None);
```

Comment:
> If this is `None`, then I don't think incremental snapshots will be created until the next full snapshot (default interval is 25,000 blocks).

Could we someshow send a snapshot request and trigger a full snapshot archive immediately after a bank_from_snapshot_dir() call?

Reply:
> If this is `None`, then I don't think incremental snapshots will be created until the next full snapshot (default interval is 25,000 blocks).

Will this be ok?  Is there any need to have a full snapshot archive present sooner than 25000 blocks?

Reply:
I am looking into the ABS code to find out how and when the hash is calculated.  Without knowing enough, my current gut feeling is that we could let ABS process the pre-stage snapshot dir into a post-stage one (i assume that means calculate hash).  
Doing this for every snapshot dir probably is too costly.  The way to reduce it probably is to set an interval.  then we won't construct a bank from a more recent pre snapshot dir, but an older post snapshot dir.  the interval need to be set to balance the tradeoffs.  Another possibility is to calculate hash as prev_hash + hash(diff).  is that doable?  If the hash only has to be done on the diff part, and the previous hash can be used, then it will be light load.

Reply:
> getting the accounts hash from the bank snapshot directly, since it's already there, and not from a new separate file.

Could you please clarify how we can get the accounts hash directly from the bank snapshot?  Do you mean deserializing the snapshot file into a data structure and read the hash field?   that snapshot file has to be a post one, not the pre one, right?

Reply:
I did some test run and see that accounts_hash_verifier calculates hash for every snapshot (every ~100 blocks apart).  It operates on the tmp directory because it gets the tmpdir from the input AccountsPackage, which is generated by the ABS.  

So, I think it is possible to do this in-place, in a pipelined stages model.

ABS generates a snapshot_dir in PRE stage.
It generates a AccountPackage not using tmpdir but the snapshot_dir itself.
AHV calculates hash, reserializes it, and marks the snapshot_dir at POST stage.
Archiving code (which thread?) archives the snapshot dir into an archive file, and mark the snapshot_dir at ARCHIVED stage.
The snapshot purging code (ABS thread) can purge any snapshot dirs at ARCHIVED stage except the the latest one.

When booting up, the latest snapshot dir at POST or ARCHIVED is good for constructing the bank.  The hash extracted is StartingSnapshotHashes.

Reply:
AccountsPackage::new_for_snapshot
"// Hard link the snapshot into a tmpdir, to ensure its not removed prior to packaging."

The purging function can check the snapshot dir stage info first before removing it, so we don't have to worry about it being removed prior to packaging.

Reply:
> For ABS & AHV, the accounts hash from POST (or ARCHIVED) will work (I believe).

If we are not sure, we can just use the sanpshot dir at the ARCHIVED stage for bank construction.  The incremental archive is done also at 100-blocks intervals.  A snapshot reaches the ARCHIVED stage shortly after the POST stage.

Reply:
I think the above mentioned pipeline stages model is a clean way.  Multiple threads do different processing (creating, computing hash, archiving, purging) in-place on the same snapshot dir, following and advancing the stages of the snapshot dir.

We could go forward in that direction.  Or, another option is simply not changing the existing design, but just patching the needed info into the snapshot dir.  When SnapshotPackagerService is done archiving so the snapshot hash (epho included), snapshot file are available for that slot, it is enough to simply copy them into the snapshot dir, making that snapshot_dir good for booting.

Not sure yet which way is better.

Reply:
I see now that StartingSnapshotHashes is not account hash, and it includes both full snapshot hash and incremental hash.  So a full snapshot hash is necessary, which won't be available at the next 100-block incremental archive time.

Then there is not many option then.  Either start generating the full archive right after booting, or let it be None till the next full archive time (up to 25000 blocks, or 3 hours).  Will that be acceptable?

Reply:
> For AccountsHashVerifier/AccountsBackgroundService, the POST bank snapshot should be sufficient for the slot and accounts hash.

Created https://buildkite.com/solana-labs/solana/builds/92960 to let AHV process the snapshot dir in-place and generates the accounts hash.

Reply:
> For SnapshotPackagerService, it should be sufficient to get the latest snapshot archives and parse their filenames to get slots and snapshot hashes.

Updated the code to do that.